### PR TITLE
Patch qsa matches closest to accept `:popover-open`

### PR DESCRIPTION
**Pull request by keithamus**
A potentially controversial change.

By default running something like `el.matches(':open')` on a browser without native Popover will result in an exception:

Chrome:
```
DOMException: Failed to execute 'matches' on 'Element': ':open' is not a valid selector.
```

Firefox:
```
DOMException: Element.matches: ':open' is not a valid selector
```

This means users of this polyfill need to work around this; either by using a `try`/`catch`, or by doing feature detection up front, e.g.:

```js
let openSelector = ':open'
try {
  document.querySelector(openSelector)
} catch {
  // use polyfilled selector instead
  openSelector = '.\\:open'
}
```

This is a little cumbersome and has to be done for every selector you intend to use, and given we're patching prototype methods anyway, perhaps it could be useful for us to patch the CSS selector functions (I believe I have them all, `querySelector`, `querySelectorAll`, `matches`, `closest`).

This change patches all of those functions taking the selector string given and replacing instances of `:open` with `.\\:open` and `:closed` with `:not(.\\:open)`. This allows us to use the `:open`/`:closed` selector and have it work with the polyfill.

## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

Tests have been added to ensure qsa/matches/closed work but it doesn't do anything close to exhaustive input testing. I've manually tested the replacer function with various inputs. We could add some unit tests to ensure its robustness, but here's how I validated it:

```js
const nonEscapedOpenSelector = /(^|[^\\]):open\b/g;
const nonEscapedClosedSelector = /(^|[^\\]):closed\b/g;
  function rewriteSelector(selector) {
    return selector
      .replace(nonEscapedOpenSelector, '$1.\\:open')
      .replace(nonEscapedClosedSelector, '$1:not(.\\:open)');
  }
console.assert(rewriteSelector(':open') == '.\\:open')
console.assert(rewriteSelector('.\\:open') == '.\\:open')
console.assert(rewriteSelector('[bar]:open') == '[bar].\\:open')
console.assert(rewriteSelector('[bar]:opened') == '[bar]:opened')
console.assert(rewriteSelector('[bar]:open.\\:open:open') == '[bar].\\:open.\\:open.\\:open')
console.assert(rewriteSelector(':closed') == ':not(.\\:open)')
console.assert(rewriteSelector('.\\:closed') == '.\\:closed')
```

## Show me
_Provide screenshots/animated gifs/videos if necessary._

# REMEMBER: Attach this PR to the Trello card

### CODE REVIEW

File: src/popover.ts [hunk 1]
```diff
@@ -24,10 +24,49 @@ function patchAttachShadow(callback: (shadowRoot: ShadowRoot) => void) {
...
+  }
+
+  patchSelectorFn(HTMLDocument.prototype, 'querySelector', rewriteSelector);
+  patchSelectorFn(HTMLDocument.prototype, 'querySelectorAll', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'querySelector', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'querySelectorAll', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'matches', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'closest', rewriteSelector);
+  patchSelectorFn(
+    DocumentFragment.prototype,
```
Comment (keithamus) [inline src/popover.ts:60]: `ShadowRoot` inherits `DocumentFragment`, and so patching `DocumentFragment` also patches `ShadowRoot`.

File: src/popover.ts [hunk 2]
```diff
@@ -24,10 +24,49 @@ function patchAttachShadow(callback: (shadowRoot: ShadowRoot) => void) {
...
+      .replace(nonEscapedOpenSelector, '$1.\\:open')
+      .replace(nonEscapedClosedSelector, '$1:not(.\\:open)');
+  }
+
+  patchSelectorFn(HTMLDocument.prototype, 'querySelector', rewriteSelector);
+  patchSelectorFn(HTMLDocument.prototype, 'querySelectorAll', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'querySelector', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'querySelectorAll', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'matches', rewriteSelector);
+  patchSelectorFn(Element.prototype, 'closest', rewriteSelector);
```
Comment (keithamus) [inline src/popover.ts:58]: `HTMLDocument` nor `DocumentFragment` have `matches` nor `closest`. I think I have all functions that take CSS selectors here. Let me know if I missed one!

File: src/popover.ts [hunk 3]
```diff
@@ -24,10 +24,49 @@ function patchAttachShadow(callback: (shadowRoot: ShadowRoot) => void) {
...
   let lastFocusedElement: HTMLElement | null = null;
 
+  function rewriteSelector(selector: string) {
+    return selector
+      .replace(nonEscapedOpenSelector, '$1.\\:open')
+      .replace(nonEscapedClosedSelector, '$1:not(.\\:open)');
+  }
+
+  patchSelectorFn(HTMLDocument.prototype, 'querySelector', rewriteSelector);
+  patchSelectorFn(HTMLDocument.prototype, 'querySelectorAll', rewriteSelector);
```
Comment (keithamus) [inline src/popover.ts:54]: I realised this should be `Document`

```suggestion
  patchSelectorFn(Document.prototype, 'querySelector', rewriteSelector);
  patchSelectorFn(Document.prototype, 'querySelectorAll', rewriteSelector);
```

File: src/popover.ts [hunk 4]
```diff
@@ -24,10 +24,49 @@ function patchAttachShadow(callback: (shadowRoot: ShadowRoot) => void) {
...
+
+const nonEscapedOpenSelector = /(^|[^\\]):open\b/g;
+const nonEscapedClosedSelector = /(^|[^\\]):closed\b/g;
+
 export function apply() {
   const visibleElements = new WeakSet<HTMLElement>();
   let lastFocusedElement: HTMLElement | null = null;
 
+  function rewriteSelector(selector: string) {
+    return selector
```
Comment (keithamus) [inline src/popover.ts:48]: Adding an early return here will speed the unhappy path up a bit:

```suggestion
    if (!selector.includes(':open')) {
      return selector;
    }
    return selector
```

File: src/popover.ts [hunk 5]
```diff
@@ -24,10 +24,52 @@ function patchAttachShadow(callback: (shadowRoot: ShadowRoot) => void) {
...
+
+const nonEscapedOpenSelector = /(^|[^\\]):open\b/g;
+const nonEscapedClosedSelector = /(^|[^\\]):closed\b/g;
+
 export function apply() {
   const visibleElements = new WeakSet<HTMLElement>();
   let lastFocusedElement: HTMLElement | null = null;
 
+  function rewriteSelector(selector: string) {
+    if (!selector.includes(':open')) {
```
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply src/popover.ts:48]: I think that's causing the failing tests too...

[the remark this answers is not in the document]
Comment (keithamus) [reply src/popover.ts:48]: Yes! :man_facepalming:

**Review by jgerigmeyer — APPROVED** [review]

**Comment by keithamus** [comment]
Looks like https://github.com/w3c/csswg-drafts/issues/8637 is moving forward with `:popover`. We should probably wait for https://github.com/whatwg/html/pull/9077 to land before we merge this.

**Comment by keithamus** [comment]
Upstream PRs have landed. `:popover-open` is the new selector. We should consider re-reviewing this PR now.

File: src/popover.css [hunk 6]
```diff
@@ -14,12 +14,22 @@
 
 /* stylelint-disable selector-class-pattern */
 @supports not selector([popover]:open) {
```
Comment (jgerigmeyer) [inline src/popover.css:16]: Is it important to retain this older `:open` selector as well? I assume just temporarily, then we might remove it?

**Review by jgerigmeyer — APPROVED** [review]

**Comment by jgerigmeyer** [comment]
Add note to changelog?

File: src/popover.css (hunk 6 above)
Comment (keithamus) [reply src/popover.css:16]: Yeah exactly

**Review by jgerigmeyer — APPROVED** [review]

**Pull request merged**